### PR TITLE
WIP: support multi-homed pod using OVN k8s CNI

### PR DESCRIPTION
**Pull request by cathy-zhou**
allow OVN k8s CNI attaching multiple network interfaces to pods.

Signed-off-by: Yun Zhou <yunz@nvidia.com>

**- What this PR does and why is it needed**

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

**Comment by girishmg** [comment]
We are going to publish the design document to capture all the information as to what is being done here. FYI @dcbw @trozet

File: go-controller/cmd/ovnkube/ovnkube.go [hunk 1]
```diff
@@ -247,8 +247,12 @@ func runOvnKube(ctx *cli.Context) error {
 		// since we capture some metrics in Start()
 		metrics.RegisterMasterMetrics(ovnNBClient, ovnSBClient)
 
-		ovnController := ovn.NewOvnController(ovnClientset, masterWatchFactory, stopChan, nil, ovnNBClient, ovnSBClient, util.EventRecorder(ovnClientset.KubeClient))
-		if err := ovnController.Start(master, wg); err != nil {
+		ovnMHController := ovn.NewOvnMHController(ovnClientset, master, masterWatchFactory,
+			stopChan, ovnNBClient, ovnSBClient, util.EventRecorder(ovnClientset.KubeClient), wg)
+		if err == nil {
```
Comment (aojea) [inline go-controller/cmd/ovnkube/ovnkube.go:252]: where is this `err` coming?

File: go-controller/pkg/cni/cnishim.go [hunk 2]
```diff
@@ -150,6 +150,12 @@ func (p *Plugin) CmdAdd(args *skel.CmdArgs) error {
 	}
 	setupLogging(conf)
 
+	if conf.NotDefault {
+		if conf.NetCidr == "" {
```
Comment (aojea) [inline go-controller/pkg/cni/cnishim.go:154]:
```suggestion
	if conf.NotDefault  &&  conf.NetCidr == "" {
```

File: go-controller/pkg/ovn/pods.go [hunk 3]
```diff
@@ -232,25 +318,58 @@ func (oc *Controller) getRoutingPodGWs(ns string) map[string][]net.IP {
...
+		// the pod is not attached to this specific network
+		klog.V(5).Infof("Pod %s/%s is not attached on this network: %s", pod.Namespace, pod.Name, oc.netconf.Name)
+		return nil
+	}
+
 	// Keep track of how long syncs take.
 	start := time.Now()
 	defer func() {
-		klog.Infof("[%s/%s] addLogicalPort took %v", pod.Namespace, pod.Name, time.Since(start))
+		klog.Infof("[%s/%s] addLogicalPort fot network %s took %v", pod.Namespace, pod.Name, oc.netconf.Name, time.Since(start))
```
Comment (aojea) [inline go-controller/pkg/ovn/pods.go:362]: nit s/fot/for/

File: go-controller/pkg/util/pod_annotation.go [hunk 4]
```diff
@@ -247,3 +262,40 @@ func GetAllPodIPs(pod *v1.Pod) ([]net.IP, error) {
...
+	if err != nil {
+		return nil, fmt.Errorf("GetK8sPodDefaultNetwork: failed to parse CRD object: %v", err)
+	}
+	if len(networks) > 1 {
+		return nil, fmt.Errorf("GetK8sPodDefaultNetwork: more than one default network is specified: %s", netAnnot)
+	}
+
+	for _, v := range networks {
+		return v, nil
+	}
```
Comment (aojea) [inline go-controller/pkg/util/pod_annotation.go:286]: at this point networks has len  0 or 1, why do you iterate?

File: go-controller/pkg/util/subnet_annotations.go [hunk 5]
```diff
@@ -43,21 +44,62 @@ const (
 	ovnNodeLocalNatIP = "k8s.ovn.org/node-local-nat-ip"
 )
 
+func updateSubnetAnnotation(annotations map[string]string, annotationName, netName string, hostSubnets []*net.IPNet) error {
```
Comment (aojea) [inline go-controller/pkg/util/subnet_annotations.go:47]: this function is a bit complex to follow, appreciate if you can add a comment explaining what is doing

Comment (cathy-zhou) [reply go-controller/pkg/util/subnet_annotations.go:50]: @aojea Thanks for reviewing this. I've addressed your comments and repushed.  At this time, the changes is not fully ready for review though, will let you know once it is ready.

**Comment by maiqueb** [comment]
Hi @cathy-zhou, @girishmg.

We - [KubeVirt](http://kubevirt.io) - are very keen for this feature. Could you share your plans for this PR ? 

Is there anything at all that I can help you with to move things forward ? 

To give you a little bit more context: what we're actually looking for is for flat L2 overlays, optionally connected to the gateway routers. Never-the-less, this PR presently does a lot of what we need for that. IMO, once this gets in, we can adapt it to also accommodate our use case.

**Pull request closed without merging**